`src/predictor.js` changes

dizmac opened the pull request:
Added `predictor#get_rating_group` and modified `predictor#render_rating_deltas` to display rating group changes.

Rating group changes are displayed if:

1. The user did not have a rating previously and the user wrote a rated contest
2. The user will get promoted/demoted after the rating change

Rating group changes will not be displayed if:

1. The user will not get promoted/demoted after the rating change

dizmac commented:
Sorry for the mess, still new to this :)

Ninjaclasher commented on src/js/predictor.js:
```diff
@@ -21,15 +21,31 @@ function on_finish_load_ratings() {
         let [delta, delta_class] = ['', 'delta-none'];
-        if (user in users && users[user].rating_change !== null) {
-            delta = users[user].rating_change;
+        let userObject = users[user];
```
```suggestion
        let user_object = users[user];
```
For consistency with the rest of the code base, let's stick to snake case.

dizmac replied:
Resolved [on commit 4da595f](https://github.com/Ninjaclasher/dmoj-rating-predictor/pull/3/commits/4da595fa3b2785b1140780f49fc46730cbdd11c3)

Ninjaclasher commented on src/js/predictor.js:
```diff
@@ -21,15 +21,31 @@ function on_finish_load_ratings() {
             delta_class = 'delta-' + deltas[Math.sign(delta)];
+        
+            if (userObject.old_rating === null || get_rating_group(userObject.old_rating) !== get_rating_group(userObject.new_rating)) {
+                $(`#user-${user} .user-name`).empty().append(`<span class='rating rate-${get_rating_group(userObject.old_rating)} user'><a href='/users/${user}'>${user}</a></span><span> → </span><span class='rating rate-${get_rating_group(userObject.new_rating)} user'><a href='/users/${user}'>${user}</a></span>`);
```
Nit: use double quotes inside HTML tags.
```suggestion
                $(`#user-${user} .user-name`).empty().append(`<span class="rating rate-${get_rating_group(userObject.old_rating)} user"><a href="/users/${user}">${user}</a></span><span> → </span><span class="rating rate-${get_rating_group(userObject.new_rating)} user"><a href="/users/${user}">${user}</a></span>`);
```
Also, let's break up the line so that it's < 120 characters for readability.

Final thing: let's use the unicode code point rather than the actual symbol (i.e. `\u2192` instead of `→`).

dizmac replied:
Resolved [on commit 4da595f](https://github.com/Ninjaclasher/dmoj-rating-predictor/pull/3/commits/4da595fa3b2785b1140780f49fc46730cbdd11c3)

Ninjaclasher reviewed (commented):
Otherwise LGTM, thanks!

dizmac commented:
Code revisions [on commit 4da595f](https://github.com/Ninjaclasher/dmoj-rating-predictor/pull/3/commits/4da595fa3b2785b1140780f49fc46730cbdd11c3)

dizmac commented:
Uh oh, I did something stupid...
Suppose to be `user` not `users`
```html
<a href='/users/${user}'>${user}</a></span>
```

Ninjaclasher commented on src/js/predictor.js:
```diff
@@ -21,15 +21,37 @@ function on_finish_load_ratings() {
+            delta = user_object.rating_change;
             delta_class = 'delta-' + deltas[Math.sign(delta)];
+        
+            if (user_object.old_rating === null || get_rating_group(user_object.old_rating) !== get_rating_group(user_object.new_rating))
```
```suggestion
            if (user_object.old_rating === null || get_rating_group(user_object.old_rating) !== get_rating_group(user_object.new_rating)) {
```
Also, please add the closing parentheses.

Ninjaclasher commented on src/js/predictor.js:
```diff
@@ -21,15 +21,37 @@ function on_finish_load_ratings() {
+            if (user_object.old_rating === null || get_rating_group(user_object.old_rating) !== get_rating_group(user_object.new_rating))
+                $(`#user-${user} .user-name`).empty()
+                    .append(
+                        `<span class='rating rate-${get_rating_group(user_object.old_rating)} user'>`+
```
Double quotes for HTML tags please.

Ninjaclasher reviewed (commented):
Just some minor nits, and we should be good.

dizmac commented:
Hopefully, we are fine now.

Ninjaclasher reviewed (commented):
I just tested this PR, and I found a few issues I didn't catch before:
 1. By default, DMOJ's username is the current rating rank, not the one at the time of contest. Right now, you're replacing both rating ranks only if it changes. However, this causes some disparity as some users will have rating ranks at the time of contest (if their ranks changed) while others will have their current rating rank (if their ranks didn't change). My suggestion would be to replace the rating rank unconditionally and show the `-> <new rating rank username>` when the rank changes. This means that all users will have rating ranks at the time of contest.
 2. For admins, the `.user-name` td has other spans as well. For example:
![image](https://user-images.githubusercontent.com/29607503/155898238-99ff7774-8687-4d13-bcf8-41d20a934d2f.png)
Using `.empty()` removes the buttons as seen in the image. I think we should only remove the `.rating` class span instead.

dizmac commented:
I implemented your first point and fixed the code based on your second point @ https://github.com/Ninjaclasher/dmoj-rating-predictor/pull/3/commits/311fae77f1cc6f004cb92d28c5dee10933d92f0a.

Ninjaclasher reviewed (approved):
LGTM, thanks!

dizmac commented:
Awesome, thanks.
